parachain/runtime/kintsugi/src/dex.rs

```diff
@@ -0,0 +1,246 @@
+        }
+        true
+    }
+}
+
+impl dex_stable::Config for Runtime {
+    type RuntimeEvent = RuntimeEvent;
+    type CurrencyId = CurrencyId;
+    type MultiCurrency = Tokens;
+    type PoolId = StablePoolId;
+    type TimeProvider = Timestamp;
+    type EnsurePoolAsset = StableAmmVerifyPoolAsset;
+    type LpGenerate = PoolLpGenerate;
+    type PoolCurrencySymbolLimit = StringLimit;
+    type PalletId = DexStablePalletId;
+    type WeightInfo = ();
+}
+
+impl dex_swap_router::Config for Runtime {
+    type RuntimeEvent = RuntimeEvent;
+    type StablePoolId = StablePoolId;
+    type Balance = Balance;
+    type CurrencyId = CurrencyId;
+    type NormalAmm = DexGeneral;
+    type StableAMM = DexStable;
+    type WeightInfo = ();
+}
+
+pub struct DexSetup;
+
+fn create_pair_and_update_rewards(asset_0: CurrencyId, asset_1: CurrencyId, rewards: Balance, fee_rate: Balance) {
+    if let Err(err) = DexGeneral::create_pair(RuntimeOrigin::root(), asset_0, asset_1, fee_rate) {
+        log::error!("Could not create pair: {:?}", err);
+    } else {
+        if let Err(err) = Farming::update_reward_schedule(
+            RuntimeOrigin::root(),
+            CurrencyId::join_lp_token(asset_0, asset_1).expect("currencies are valid; qed"),
+            CurrencyId::Token(KINT),
+            // period is 5 (blocks per minute)
+            // so period count for three months
+            // is 60 minutes * 24 hours * 92 days
+            132480,
+            rewards,
+        ) {
+            log::error!("Could not update rewards: {:?}", err);
+        }
+    }
+}
+
+impl OnRuntimeUpgrade for DexSetup {
+    fn on_runtime_upgrade() -> Weight {
+        create_pair_and_update_rewards(
+            CurrencyId::Token(KBTC),
+            CurrencyId::Token(KSM),
+            45000000000000000, // 45,000 KINT
+            15,                // 0.15%
+        );
+
+        create_pair_and_update_rewards(
+            CurrencyId::Token(KBTC),
+            CurrencyId::ForeignAsset(3), // USDT
+            40000000000000000,           // 40,000 KINT
+            15,                          // 0.15%
+        );
+
+        create_pair_and_update_rewards(
+            CurrencyId::Token(KSM),
+            CurrencyId::Token(KINT),
+            35000000000000000, // 35,000 KINT
+            25,                // 0.25%
+        );
+
+        Default::default()
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+        ensure!(
+            dex_general::PairStatuses::<Runtime>::iter().collect::<Vec<_>>().len() == 0,
+            "Should not have pools"
+        );
+        Ok(Vec::new())
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn post_upgrade(_state: Vec<u8>) -> Result<(), &'static str> {
+        ensure!(
+            dex_general::PairStatuses::<Runtime>::iter().collect::<Vec<_>>().len() == 3,
+            "Should have pools"
+        );
+        Ok(())
+    }
+}
+
+pub struct LoansSetup;
```

Comment:
Any reason to keep `LoansSetup` in `dex.rs` instead of its own file?

Reply:
We will remove it as soon as it has run on Kintsugi.